src/GenericCyclotomics.jl

```diff
@@ -0,0 +1,445 @@
+using Oscar.Random: Random, SamplerTrivial, GLOBAL_RNG
+using Oscar.RandomExtensions: RandomExtensions, Make2, AbstractRNG
+
+import Oscar.AbstractAlgebra: parent_type, elem_type, base_ring, base_ring_type, parent, is_domain_type,
+	is_exact_type, canonical_unit, isequal, divexact, zero!, mul!, add!, addeq!,
+	get_cached!, is_unit, characteristic, Ring, RingElem, expressify, evaluate
+import Oscar: pretty, Indent, Dedent, terse, is_terse
+import Oscar.AbstractAlgebra: normal_form
+
+import Base: show, +, -, *, ^, ==, inv, isone, iszero, one, zero, rand, deepcopy_internal, hash
+
+const UPoly = Generic.UnivPoly{QQFieldElem, Generic.MPoly{QQFieldElem}}
+const UPolyRing = parent_type(UPoly)
+const UPolyFrac = Generic.FracFieldElem{UPoly}
+const UPolyFracRing = Generic.FracField{UPoly}
+
+"""
+    normal_form(f::AbstractAlgebra.Generic.UnivPoly{ZZRingElem, AbstractAlgebra.Generic.MPoly{ZZRingElem}}, m::Int64)
+
+Return a normal form of `f` modulo `m`, such that `normal_form(f,m)` is equal
+to `normal_form(g,m)` if and only if `f` and `g` are congruent modulo `m`.
+
+# Examples
+```jldoctest
+julia> R=universal_polynomial_ring(ZZ);
+
+julia> x=gen(R, :x);
+
+julia> normal_form(4*x^2,6)
+x^2 + 3*x
+
+julia> normal_form(4*x^2-x^2-3*x,6)
+0
+
+```
+"""
+function normal_form(f::AbstractAlgebra.Generic.UnivPoly{ZZRingElem, AbstractAlgebra.Generic.MPoly{ZZRingElem}}, m::Int64)
+	if m < 1
+		throw(DomainError(m, "A normal form for non-positive moduli is not defined!"))
+	end
+	R=parent(f)
+	if is_constant(f)
+		return R(mod(constant_coefficient(f),m))
+	end
+	# find the variable with highest index occurring in f
+	i=findlast(x -> degree(f,x)>0, gens(R))
+	x_i=R[i]
+
+	# factor m for later use
+	factorization=factor(m)
+
+	# compute m-th Kempner number by using the schemes introduced by Kempner
+	# these schemes can later be reused to compute reductions of m
```

Comment:
Please add references! I.e. add the Kempner paper to the bib and then point to specific sections.

That would help anyone else trying to understand this code.